src/main/c/NativeModule.cpp

```diff
@@ -0,0 +1,625 @@
+    write_mak();
+    write_cpp();
+    std::stringstream ss;
+    ss << "make -B -C " << config.module_dir_ << " -f " << hm_mak_ << " 1>/dev/null";
+    // run the command synchronously, while holding the big_mutex
+    int rc = system(ss.str().c_str());
+    return (rc == 0);
+  }
+};
+
+NativeModule::NativeModule(
+  const char* options,
+  const char* source,
+  const char* include,
+  bool force_build
+) :
+  build_state_(kInit),
+  load_state_(kInit),
+  key_(hash_two_strings(options, source)),
+  is_global_(false),
+  dlopen_handle_(nullptr),
+  lib_name_(config.get_lib_name(key_)),
+  new_name_(config.get_new_name(key_)) {
+  std::lock_guard<std::mutex> mylock(big_mutex);
+  // Master constructor - try to get module built in local file
+  config.ensure_module_dir_exists();
+  bool have_lib = (!force_build && file_exists(lib_name_));
+  if (have_lib) {
+    build_state_ = kPass;
+  } else {
+    // The file doesn't exist, let's build it
+    ModuleBuilder builder(options, source, include, key_);
+    build_state_ = (builder.try_to_build() ? kPass : kFail);
+  }
+}
+
+NativeModule::NativeModule(
+  bool is_global,
+  const char* key,
+  ssize_t binary_size,
+  const void* binary
+) :
+  build_state_(is_global ? kPass : kInit),
+  load_state_(is_global ? kPass : kInit),
+  key_(key),
+  is_global_(is_global),
+  dlopen_handle_(nullptr),
+  lib_name_(config.get_lib_name(key_)),
+  new_name_(config.get_new_name(key_)) {
+  std::lock_guard<std::mutex> mylock(big_mutex);
+  // Worker constructor - try to get the binary written to local file
+  if (is_global_) return;
+  int rc = 0;
+  config.ensure_module_dir_exists();
+  struct stat lib_stat;
+  if (file_stat(lib_name_, &lib_stat) && (lib_stat.st_size == binary_size)) {
+    build_state_ = kPass;
+  } else {
+    // We hold big_mutex so there is no race
+    int fd = open(new_name_.c_str(), O_WRONLY|O_CREAT|O_TRUNC, 0666);
+    if (fd < 0) {
+      perror("open");
+      assert(0);
+    }
+    // Now we're about to write the new file
+    rc = write(fd, binary, binary_size);
+    assert(rc == binary_size);
+    ::close(fd);
+    ::chmod(new_name_.c_str(), 0644);
+    struct stat st;
+    if (file_stat(lib_name_, &st)) {
+      long old_size = st.st_size;
+      if (old_size == binary_size) {
+        ::unlink(new_name_.c_str());
+      } else if (old_size == 0) {
+        ::unlink(lib_name_.c_str());
+      } else {
+        auto old = lib_name_ + ".old";
+        ::rename(lib_name_.c_str(), old.c_str());
+      }
+      // Don't let anyone see the file until it is completely written
+      rc = ::rename(new_name_.c_str(), lib_name_.c_str());
+      build_state_ = ((rc == 0) ? kPass : kFail);
+    }
+  }
+  if (build_state_ == kPass) {
+    try_load_locked();
+  }
+}
+
+NativeModule::~NativeModule() {
+  if (!is_global_ && dlopen_handle_) {
+    dlclose(dlopen_handle_);
+  }
+}
+
+bool NativeModule::try_wait_for_build_locked() {
+  if (build_state_ == kInit) {
```

Comment:
This function seems mis-named now.  I would assume build_state now means "does the library file exist", in which case either constructor should set it to kPass or kFail (depending on whether the build or writing the binary file succeeded or not) and this should never happen.